File: examples/word-counts/segment_word_counts.py
import logging
import os
import pickle
import tempfile

import tornado.ioloop

from pyworker.tornado import MessageQueueDispatcher


def word_counts(correlation_id: str, msg):
    job_id = correlation_id.split(':', maxsplit=1)[0]
    filename = msg['filename']
    section = msg['section']
    word_counts = {}
    with open(filename, 'r') as fp:
        lineno = 0
        while True:
            line = fp.readline()
            if not line:
                break
            if lineno == section['end']:
                break

            if lineno >= section['start']:
                tokens = line.split()
                for tok in tokens:
                    word_counts[tok] = word_counts.get(tok, 0) + 1

            lineno += 1

    output = os.path.join(tempfile.gettempdir(), job_id + '_counts.bin')
    with open(output, 'wb') as fp:
        pickle.dump(word_counts, fp)

    # task end gathers results from all jobs
    return dict(section=output)


def main():
    logger = logging.getLogger('MessageQueueDispatcher')
    logger.setLevel(logging.DEBUG)
    dispatcher = MessageQueueDispatcher(
        os.environ['AMQP_SERVER'], "segment-word-counts", word_counts
    )
    dispatcher.connect()

    ioloop = tornado.ioloop.IOLoop.current()
    ioloop.start()


if __name__ == '__main__':
    main()

File: examples/word-counts-js/src/segment_word_counts.ts
import * as fs from 'fs';
import * as readline from 'readline';
import * as os from 'os';
import * as path from 'path';
import QueueListener from 'jsworker';

interface Section {
  start: number;
  end: number;
}
interface IMessage {
  filename: string;
  section: Section;
}

interface IResult {
  section: string;
}

class Deferred<T> {
  readonly promise: Promise<T>;
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  resolve = (value: T) => {};
  // eslint-disable-next-line @typescript-eslint/no-explicit-any,@typescript-eslint/no-unused-vars
  reject = (value: any) => {};
  constructor() {
    this.promise = new Promise<T>((resolve, reject) => {
      this.resolve = resolve;
      this.reject = reject;
    });
  }
}

async function wordCounts(
  correlationId: string,
  msg: IMessage
): Promise<IResult> {
  const reader = readline.createInterface(fs.createReadStream(msg.filename));
  const counts = new Map<string, number>();
  let lineno = 0;

  reader.on('line', line => {
    const accept = lineno >= msg.section.start && lineno < msg.section.end;
    if (accept) {
      const words = line.split(/\s+/);
      words.forEach(w => {
        if (!w) {
          return;
        }
        const value = counts.has(w) ? counts.get(w) : 0;
        counts.set(w, value! + 1);
      });
    }
    lineno = lineno + 1;
  });

  const jobId = correlationId.split(':')[0];
  const output = path.join(os.tmpdir(), jobId + '_s_counts.tsv');

  const deferred = new Deferred<IResult>();

  reader.on('close', () => {
    let data = '';
    counts.forEach((value, key) => {
      data += `${key}\t${value}\n`;
    });
    fs.writeFile(output, data, err => {
      if (err) {
        throw err;
      }
      deferred.resolve({section: output});
    });
  });

  return deferred.promise;
}

function main() {
  const amqpServer = process.env.AMQP_SERVER;
  if (!amqpServer) {
    console.log('No AMQP_SERVER env variable not defined');
  }
  const listener = new QueueListener(
    amqpServer!,
    'segment-word-counts',
    wordCounts
  );
  listener.connect();
}

main();
